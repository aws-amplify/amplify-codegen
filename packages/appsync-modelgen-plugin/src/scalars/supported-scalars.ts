//  Used only in  tests. Directive definition will be passed as part of the configuration when modelgen is run using CLI
export const scalars = [
  'ID',
  'String',
  'Int',
  'Float',
  'Boolean',
  'AWSDate',
  'AWSDateTime',
  'AWSTime',
  'AWSTimestamp',
  'AWSEmail',
  'AWSJSON',
  'AWSURL',
  'AWSPhone',
  'AWSIPAddress',
]
  .map(typeName => `scalar ${typeName}`)
  .join();
