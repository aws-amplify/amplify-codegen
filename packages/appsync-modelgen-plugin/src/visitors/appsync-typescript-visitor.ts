import { indentMultiline } from '@graphql-codegen/visitor-plugin-common';
import { TypeScriptDeclarationBlock } from '../languages/typescript-declaration-block';
import {
  AppSyncModelVisitor,
  CodeGenEnum,
  CodeGenField,
  CodeGenModel,
  CodeGenPrimaryKeyType,
  ParsedAppSyncModelConfig,
  RawAppSyncModelConfig,
} from './appsync-visitor';

export interface RawAppSyncModelTypeScriptConfig extends RawAppSyncModelConfig {}
export interface ParsedAppSyncModelTypeScriptConfig extends ParsedAppSyncModelConfig {
  isDeclaration: boolean;
}

export class AppSyncModelTypeScriptVisitor<
  TRawConfig extends RawAppSyncModelTypeScriptConfig = RawAppSyncModelTypeScriptConfig,
  TPluginConfig extends ParsedAppSyncModelTypeScriptConfig = ParsedAppSyncModelTypeScriptConfig
> extends AppSyncModelVisitor<TRawConfig, TPluginConfig> {
  protected SCALAR_TYPE_MAP: { [key: string]: string } = {
    String: 'string',
    Int: 'number',
    Float: 'number',
    Boolean: 'boolean',
    ID: 'string',
  };

  protected IMPORT_STATEMENTS = [
    'import { ModelInit, MutableModel, PersistentModelConstructor } from "@aws-amplify/datastore";',
    'import { initSchema } from "@aws-amplify/datastore";',
    '',
    'import { schema } from "./schema";',
  ];

  protected BASE_DATASTORE_IMPORT = new Set(['ModelInit', 'MutableModel', 'PersistentModelConstructor', '__modelMeta__']);

  protected MODEL_META_FIELD_NAME = '__modelMeta__';

  generate(): string {
    // TODO: Remove us, leaving in to be explicit on why this flag is here.
    const shouldUseModelNameFieldInHasManyAndBelongsTo = false;
    this.processDirectives(shouldUseModelNameFieldInHasManyAndBelongsTo);
    const imports = this.generateImports();
    const enumDeclarations = Object.values(this.enumMap)
      .map(enumObj => this.generateEnumDeclarations(enumObj))
      .join('\n\n');

    const modelDeclarations = Object.values(this.modelMap)
      .map(typeObj => this.generateModelDeclaration(typeObj))
      .join('\n\n');

    const nonModelDeclarations = Object.values(this.nonModelMap)
      .map(typeObj => this.generateModelDeclaration(typeObj, true, false))
      .join('\n\n');

    const modelInitialization = this.generateModelInitialization([...Object.values(this.modelMap), ...Object.values(this.nonModelMap)]);

    const modelExports = this.generateExports(Object.values(this.modelMap));

    return [imports, enumDeclarations, modelDeclarations, nonModelDeclarations, modelInitialization, modelExports].join('\n\n');
  }

  protected generateImports(): string {
    return this.IMPORT_STATEMENTS.join('\n');
  }

  protected generateEnumDeclarations(enumObj: CodeGenEnum, exportEnum: boolean = false): string {
    const enumDeclarations = new TypeScriptDeclarationBlock()
      .asKind('enum')
      .withName(this.getEnumName(enumObj))
      .withEnumValues(enumObj.values)
      .export(exportEnum);

    return enumDeclarations.string;
  }

  protected generateModelMetaData(modelObj: CodeGenModel): string {
    const modelName = this.generateModelTypeDeclarationName(modelObj);
    const modelDeclarations = new TypeScriptDeclarationBlock()
      .asKind('type')
      .withName(`${modelName}MetaData`)
      .export(false);

    modelDeclarations.addProperty('identifier', this.constructIdentifier(modelObj));

    const isTimestampFeatureFlagEnabled = this.config.isTimestampFieldsAdded;
    let readOnlyFieldNames: string[] = [];

    modelObj.fields.forEach((field: CodeGenField) => {
      if (isTimestampFeatureFlagEnabled && field.isReadOnly) {
        readOnlyFieldNames.push(`'${field.name}'`);
      }
    });
    if (readOnlyFieldNames.length) {
      modelDeclarations.addProperty('readOnlyFields', readOnlyFieldNames.join(' | '));
    }

    return modelDeclarations.string;
  }

  protected generateModelMetaDataType(modelObj: CodeGenModel): string {
    let readOnlyFieldNames: string[] = [];
    modelObj.fields.forEach((field: CodeGenField) => {
      if (field.isReadOnly) {
        readOnlyFieldNames.push(`'${field.name}'`);
      }
    });
    let modelMetaFields: string[] = [];
    //Custom primary key meta field
    modelMetaFields.push(`identifier: ${this.constructIdentifier(modelObj)};`);
    //Read only meta field
    if (readOnlyFieldNames.length) {
      modelMetaFields.push(`readOnlyFields: ${readOnlyFieldNames.join(' | ')};`);
    }
    const result = [
      '{',
      indentMultiline(modelMetaFields.join('\n')),
      '}'
    ];
    return result.join('\n');
  }

  protected constructIdentifier(modelObj: CodeGenModel): string {
    const primaryKeyField = modelObj.fields.find(f => f.primaryKeyInfo)!;
    const { primaryKeyType, sortKeyFields } = primaryKeyField.primaryKeyInfo!;
    switch (primaryKeyType) {
      case CodeGenPrimaryKeyType.ManagedId:
        this.BASE_DATASTORE_IMPORT.add('ManagedIdentifier');
        return `ManagedIdentifier<${modelObj.name}, '${primaryKeyField.name}'>`;
      case CodeGenPrimaryKeyType.OptionallyManagedId:
        this.BASE_DATASTORE_IMPORT.add('OptionallyManagedIdentifier');
        return `OptionallyManagedIdentifier<${modelObj.name}, '${primaryKeyField.name}'>`;
      case CodeGenPrimaryKeyType.CustomId:
        this.BASE_DATASTORE_IMPORT.add('CustomIdentifier');
        const identifierFields: string[] = [primaryKeyField.name, ...sortKeyFields].filter(f => f);
        const identifierFieldsStr = identifierFields.length === 1 ? `'${identifierFields[0]}'` : `[${identifierFields.map(fieldStr => `'${fieldStr}'`).join(', ')}]`
        return `CustomIdentifier<${modelObj.name}, ${identifierFieldsStr}>`;
    }
  }

  /**
   *
   * @param modelObj CodeGenModel object
   * @param isDeclaration flag indicates if the class needs to be exported
   */
  protected generateModelDeclaration(modelObj: CodeGenModel, isDeclaration: boolean = true, isModelType: boolean = true): string {
    const modelName = this.generateModelTypeDeclarationName(modelObj);
    const modelDeclarations = new TypeScriptDeclarationBlock()
      .asKind('class')
      .withFlag({ isDeclaration })
      .withName(modelName)
      .export(true);

    //Add model meta field
    if (isModelType) {
      modelDeclarations.addProperty(`[${this.MODEL_META_FIELD_NAME}]`, this.generateModelMetaDataType(modelObj), undefined, 'DEFAULT', {
        readonly: true,
        optional: false
      })
    }
    modelObj.fields.forEach((field: CodeGenField) => {
      modelDeclarations.addProperty(this.getFieldName(field), this.getNativeType(field), undefined, 'DEFAULT', {
        readonly: true,
        optional: field.isList ? field.isListNullable : field.isNullable,
      });
    });

    // Constructor
    modelDeclarations.addClassMethod(
      'constructor',
      null,
      null,
      [
        {
          name: 'init',
          type: `ModelInit<${modelName}>`,
        },
      ],
      'DEFAULT',
      {},
    );

    // copyOf method
    if (Object.values(this.modelMap).includes(modelObj)) {
      modelDeclarations.addClassMethod(
        'copyOf',
        modelName,
        null,
        [
          {
            name: 'source',
            type: modelName,
          },
          {
            name: 'mutator',
            type: `(draft: MutableModel<${modelName}>) => MutableModel<${modelName}> | void`,
          },
        ],
        'DEFAULT',
        { static: true },
      );
    }
    return modelDeclarations.string;
  }

  /**
   * Generate model Declaration using classCreator
   * @param model
   */
  protected generateModelInitialization(models: CodeGenModel[], includeTypeInfo: boolean = true): string {
    if (models.length === 0) {
      return '';
    }
    const modelClasses = models
      .map(model => [this.generateModelImportName(model), this.generateModelImportAlias(model)])
      .map(([importName, aliasName]) => {
        return importName === aliasName ? importName : `${importName}: ${aliasName}`;
      });

    const initializationResult = ['const', '{', modelClasses.join(', '), '}', '=', 'initSchema(schema)'];
    if (includeTypeInfo) {
      const typeInfo = models
        .map(model => {
          return [this.generateModelImportName(model), this.generateModelTypeDeclarationName(model)];
        })
        .map(([importName, modelDeclarationName]) => `${importName}: PersistentModelConstructor<${modelDeclarationName}>;`);
      const typeInfoStr = ['{', indentMultiline(typeInfo.join('\n')), '}'].join('\n');
      initializationResult.push('as', typeInfoStr);
    }
    return `${initializationResult.join(' ')};`;
  }

  protected generateExports(modelsOrEnum: (CodeGenModel | CodeGenEnum)[]): string {
    const exportStr = modelsOrEnum
      .map(model => {
        if (model.type === 'model') {
          const modelClassName = this.generateModelImportAlias(model);
          const exportClassName = this.getModelName(model);
          return modelClassName !== exportClassName ? `${modelClassName} as ${exportClassName}` : modelClassName;
        }
        return model.name;
      })
      .join(',\n');
    return ['export {', indentMultiline(exportStr), '};'].join('\n');
  }

  /**
   * Generate the type declaration class name of Model
   * @param model CodeGenModel
   */
  protected generateModelTypeDeclarationName(model: CodeGenModel): string {
    return `${this.getModelName(model)}Model`;
  }

  /**
   * Generate alias for the model used when importing it from initSchema
   * @param model
   */
  protected generateModelImportAlias(model: CodeGenModel): string {
    return this.getModelName(model);
  }

  /**
   * Generate the import name for model from initSchema
   * @param model Model object
   *
   */
  protected generateModelImportName(model: CodeGenModel): string {
    return this.getModelName(model);
  }

  /**
   * Generate the class name for export
   * @param model
   */
  protected generateModelExportName(model: CodeGenModel): string {
    return this.getModelName(model);
  }

  protected getListType(typeStr: string, field: CodeGenField): string {
    let type: string = typeStr;
    if (field.isNullable) {
      type = `(${type} | null)`;
    }
    return `${type}[]`;
  }

  protected getNativeType(field: CodeGenField): string {
    const typeName = field.type;
    const isNullable = field.isList ? field.isListNullable : field.isNullable;
    const nullableTypeUnion = isNullable ? ' | null' : '';
    if (this.isModelType(field)) {
      const modelType = this.modelMap[typeName];
      const typeNameStr = this.generateModelTypeDeclarationName(modelType);
      return (field.isList ? this.getListType(typeNameStr, field) : typeNameStr) + nullableTypeUnion;
    }

    let nativeType = super.getNativeType(field);

    if (this.isEnumType(field)) {
      nativeType = `${nativeType} | keyof typeof ${this.getEnumName(this.enumMap[typeName])}`;
    }

    nativeType = nativeType + nullableTypeUnion;

    return nativeType;
  }
}
