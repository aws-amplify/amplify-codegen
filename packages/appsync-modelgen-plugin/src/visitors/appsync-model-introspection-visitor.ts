import { DEFAULT_SCALARS, NormalizedScalarsMap } from "@graphql-codegen/visitor-plugin-common";
import { GraphQLSchema } from "graphql";
import { AssociationType, Field, Fields, FieldType, ModelAttribute, ModelIntrospectionSchema, PrimaryKeyInfo, SchemaEnum, SchemaModel, SchemaNonModel } from "../interfaces/introspection";
import { METADATA_SCALAR_MAP } from "../scalars";
import { CodeGenConnectionType } from "../utils/process-connections";
import { RawAppSyncModelConfig, ParsedAppSyncModelConfig, AppSyncModelVisitor, CodeGenEnum, CodeGenField, CodeGenModel, CodeGenPrimaryKeyType } from "./appsync-visitor";
import fs from 'fs';
import path from 'path';
import Ajv from 'ajv';

export interface RawAppSyncModelIntrospectionConfig extends RawAppSyncModelConfig {};
export interface ParsedAppSyncModelIntrospectionConfig extends ParsedAppSyncModelConfig {};
export class AppSyncModelIntrospectionVisitor<
  TRawConfig extends RawAppSyncModelIntrospectionConfig = RawAppSyncModelIntrospectionConfig,
  TPluginConfig extends ParsedAppSyncModelIntrospectionConfig = ParsedAppSyncModelIntrospectionConfig
> extends AppSyncModelVisitor<TRawConfig, TPluginConfig> {
  private readonly introspectionVersion = 1;
  private schemaValidator: Ajv.ValidateFunction;
  constructor(
    schema: GraphQLSchema,
    rawConfig: TRawConfig,
    additionalConfig: Partial<TPluginConfig>,
    defaultScalars: NormalizedScalarsMap = DEFAULT_SCALARS,
  ) {
    super(schema, rawConfig, additionalConfig, defaultScalars);
    const modelIntrospectionSchemaText = fs.readFileSync(path.join(__dirname, '..', '..', 'schemas', 'introspection', this.introspectionVersion.toString(), 'ModelIntrospectionSchema.json'), 'utf8');
    const modelIntrospectionSchema = JSON.parse(modelIntrospectionSchemaText);
    this.schemaValidator = new Ajv().compile(modelIntrospectionSchema);
  }

  generate(): string {
    const shouldUseModelNameFieldInHasManyAndBelongsTo = false;
    // This flag is going to be used to tight-trigger on JS implementations only.
    const shouldImputeKeyForUniDirectionalHasMany = true;
    const shouldUseFieldsInAssociatedWithInHasOne = true;
    this.processDirectives(
      shouldUseModelNameFieldInHasManyAndBelongsTo,
      shouldImputeKeyForUniDirectionalHasMany,
      shouldUseFieldsInAssociatedWithInHasOne
    );

    const modelIntrosepctionSchema = this.generateModelIntrospectionSchema();
    if (!this.schemaValidator(modelIntrosepctionSchema)) {
      throw new Error(`Data did not validate against the supplied schema. Underlying errors were ${JSON.stringify(this.schemaValidator.errors)}`);
    }
    return JSON.stringify(modelIntrosepctionSchema, null, 4);
  }

  protected generateModelIntrospectionSchema(): ModelIntrospectionSchema {
    const result: ModelIntrospectionSchema = {
      version: this.introspectionVersion,
      models: {},
      enums: {},
      nonModels: {},
    };

    const models = Object.values(this.getSelectedModels()).reduce((acc, model: CodeGenModel) => {
      return { ...acc, [model.name]: this.generateModelMetadata(model) };
    }, {});
    const nonModels = Object.values(this.getSelectedNonModels()).reduce((acc, nonModel: CodeGenModel) => {
      return { ...acc, [nonModel.name]: this.generateNonModelMetadata(nonModel) };
    }, {});
    const enums = Object.values(this.enumMap).reduce((acc, enumObj) => {
      return { ...acc, [this.getEnumName(enumObj)]: this.generateEnumMetadata(enumObj) };
    }, {});
    return { ...result, models, nonModels, enums };
  }

  private getFieldAssociation(field: CodeGenField): AssociationType | void {
    if (field.connectionInfo) {
      const { connectionInfo } = field;
      const connectionAttribute: any = { connectionType: connectionInfo.kind };
      if (connectionInfo.kind === CodeGenConnectionType.HAS_MANY) {
        connectionAttribute.associatedWith = connectionInfo.associatedWithFields.map(f => this.getFieldName(f));
      } else if (connectionInfo.kind === CodeGenConnectionType.HAS_ONE) {
          connectionAttribute.associatedWith = connectionInfo.associatedWithFields.map(f => this.getFieldName(f));
          connectionAttribute.targetNames = connectionInfo.targetNames;
      } else {
        connectionAttribute.targetNames = connectionInfo.targetNames;
      }
      return connectionAttribute;
    }
  }

  private generateModelAttributes(model: CodeGenModel): ModelAttribute[] {
    return model.directives.map(d => ({
      type: d.name,
      properties: d.arguments,
    }));
  }
  private generateModelMetadata(model: CodeGenModel): SchemaModel {
    return {
      ...this.generateNonModelMetadata(model),
      syncable: true,
      pluralName: this.pluralizeModelName(model),
      attributes: this.generateModelAttributes(model),
      primaryKeyInfo: this.generateModelPrimaryKeyInfo(model),
    };
  }

  private generateNonModelMetadata(nonModel: CodeGenModel): SchemaNonModel {
    return {
      name: this.getModelName(nonModel),
      fields: nonModel.fields.reduce((acc: Fields, field: CodeGenField) => {
        const fieldMeta: Field = {
          name: this.getFieldName(field),
          isArray: field.isList,
          type: this.getType(field.type),
          isRequired: !field.isNullable,
          attributes: [],
        };

        if (field.isListNullable !== undefined) {
          fieldMeta.isArrayNullable = field.isListNullable;
        }

        if (field.isReadOnly !== undefined) {
          fieldMeta.isReadOnly = field.isReadOnly;
        }

        const association: AssociationType | void = this.getFieldAssociation(field);
        if (association) {
          fieldMeta.association = association;
        }
        acc[field.name] = fieldMeta;
        return acc;
      }, {}),
    };
  }
  private generateEnumMetadata(enumObj: CodeGenEnum): SchemaEnum {
    return {
      name: enumObj.name,
      values: Object.values(enumObj.values),
    };
  }

  protected getType(gqlType: string): FieldType {
    // Todo: Handle unlisted scalars
    if (gqlType in METADATA_SCALAR_MAP) {
      return METADATA_SCALAR_MAP[gqlType] as FieldType;
    }
    if (gqlType in this.enumMap) {
      return { enum: this.enumMap[gqlType].name };
    }
    if (gqlType in this.nonModelMap) {
      return { nonModel: gqlType };
    }
    if (gqlType in this.modelMap) {
      return { model: gqlType };
    }
    throw new Error(`Unknown type ${gqlType}`);
  }

  private generateModelPrimaryKeyInfo(model: CodeGenModel): PrimaryKeyInfo {
    const primaryKeyField = this.getModelPrimaryKeyField(model);
    if (primaryKeyField && primaryKeyField.primaryKeyInfo) {
      const { primaryKeyType, sortKeyFields } = primaryKeyField.primaryKeyInfo;
      return {
        isCustomPrimaryKey: primaryKeyType === CodeGenPrimaryKeyType.CustomId,
        primaryKeyFieldName: this.getFieldName(primaryKeyField),
        sortKeyFieldNames: sortKeyFields.map(field => this.getFieldName(field))
      };
    }
    throw new Error(`No primary key found for model ${model.name}`);
  }
}