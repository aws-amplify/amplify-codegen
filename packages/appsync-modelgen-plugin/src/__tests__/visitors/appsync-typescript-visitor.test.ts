import { buildSchema, GraphQLSchema, parse, visit } from 'graphql';
import { validateTs } from '@graphql-codegen/testing';
import { DefaultDirectives } from '@aws-amplify/graphql-directives';
import { TYPESCRIPT_SCALAR_MAP } from '../../scalars';
import { scalars } from '../../scalars/supported-scalars';
import { AppSyncModelTypeScriptVisitor } from '../../visitors/appsync-typescript-visitor';

const directives = DefaultDirectives.map(directive => directive.definition).join('\n');

const buildSchemaWithDirectives = (schema: String): GraphQLSchema => {
  return buildSchema([schema, directives, scalars].join('\n'));
};
const getVisitor = (schema: string): AppSyncModelTypeScriptVisitor => {
  const ast = parse(schema);
  const builtSchema = buildSchemaWithDirectives(schema);
  const visitor = new AppSyncModelTypeScriptVisitor(
    builtSchema,
    { directives, target: 'typescript', scalars: TYPESCRIPT_SCALAR_MAP, codegenVersion: '3.3.4' },
    {},
  );
  visit(ast, { leave: visitor });
  return visitor;
};

describe('TypeScript visitor', () => {
  test('singular enum', () => {
    const schema = /* GraphQL */ `
      enum Frequency {
        YEARLY
        WEEKLY
      }

      type Recurrence {
        frequency: Frequency!
      }
    `;
    const visitor = getVisitor(schema);
    expect(visitor.generate()).toMatchSnapshot();
  });

  test('list enum', () => {
    const schema = /* GraphQL */ `
      enum DayOfWeek {
        MONDAY
        TUESDAY
        WEDNESDAY
        THURSDAY
        FRIDAY
        SATURDAY
        SUNDAY
      }

      type Recurrence {
        daysOfWeek: [DayOfWeek!]!
      }
    `;
    const visitor = getVisitor(schema);
    expect(visitor.generate()).toMatchSnapshot();
  });
});
