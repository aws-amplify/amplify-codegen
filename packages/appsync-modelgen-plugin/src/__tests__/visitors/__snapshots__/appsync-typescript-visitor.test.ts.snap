// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeScript visitor list enum 1`] = `
"import { ModelInit, MutableModel, PersistentModelConstructor } from \\"@aws-amplify/datastore\\";
import { initSchema } from \\"@aws-amplify/datastore\\";

import { schema } from \\"./schema\\";

export enum DayOfWeek {
  MONDAY = \\"MONDAY\\",
  TUESDAY = \\"TUESDAY\\",
  WEDNESDAY = \\"WEDNESDAY\\",
  THURSDAY = \\"THURSDAY\\",
  FRIDAY = \\"FRIDAY\\",
  SATURDAY = \\"SATURDAY\\",
  SUNDAY = \\"SUNDAY\\"
}



type EagerRecurrenceModel = {
  readonly daysOfWeek: DayOfWeek[] | Array<keyof typeof DayOfWeek>;
}

type LazyRecurrenceModel = {
  readonly daysOfWeek: DayOfWeek[] | Array<keyof typeof DayOfWeek>;
}

export declare type RecurrenceModel = LazyLoading extends LazyLoadingDisabled ? EagerRecurrenceModel : LazyRecurrenceModel

export declare const RecurrenceModel: (new (init: ModelInit<RecurrenceModel>) => RecurrenceModel)

const { Recurrence } = initSchema(schema) as {
  Recurrence: PersistentModelConstructor<RecurrenceModel>;
};

export {
  
};"
`;

exports[`TypeScript visitor singular enum 1`] = `
"import { ModelInit, MutableModel, PersistentModelConstructor } from \\"@aws-amplify/datastore\\";
import { initSchema } from \\"@aws-amplify/datastore\\";

import { schema } from \\"./schema\\";

export enum Frequency {
  YEARLY = \\"YEARLY\\",
  WEEKLY = \\"WEEKLY\\"
}



type EagerRecurrenceModel = {
  readonly frequency: Frequency | keyof typeof Frequency;
}

type LazyRecurrenceModel = {
  readonly frequency: Frequency | keyof typeof Frequency;
}

export declare type RecurrenceModel = LazyLoading extends LazyLoadingDisabled ? EagerRecurrenceModel : LazyRecurrenceModel

export declare const RecurrenceModel: (new (init: ModelInit<RecurrenceModel>) => RecurrenceModel)

const { Recurrence } = initSchema(schema) as {
  Recurrence: PersistentModelConstructor<RecurrenceModel>;
};

export {
  
};"
`;
